Command expansion v2

tdaron opened the pull request:
Hi !

We talked about command expansions inside of #6979, but I created a new PR as the original PR reached a "stopping" point as the PR became inactive, mainly because the OP is currently busy. 

Into this new PR , i rebased #6979 on master, and made some changes as @ksdrar told me to ;) (I think whitespaces are fixed as i handle completions inside of the shellwords parsing phase)

If you think creating a new PR is not a good idea i can of course close this one, but I think it would be great to finally get this feature ! 

Current variables: 

| Variable                | Description |
| ---                     | ---                      |
| `%{basename}` or `%{b}` | The name and extension of the currently focused file. |
| `%{dirname}`  or `%{d}` | The absolute path of the parent directory of the currently focused file. |
| `%{cwd}`                | The absolute path of the current working directory of Helix. |
| `%{repo}`               | The absolute path of the VCS repository helix is opened in. Fallback to `cwd` if not inside a VCS repository|
| `%{filename}` or `%{f}` | The absolute path of the currently focused file. |
| `%{filename:rel}`       | The relative path of the file according to `cwd` (will give absolute path if the file is not a child of the current working directory) |
| `%{filename:repo_rel}`  | The relative path of the file according to `repo` (will give absolute path if the file is not a child of the VCS directory or the cwd) |
| `%{ext}`                | The extension of the current file |
| `%{lang}`               | The language of the current file   |
| `%{linenumber}`         | The line number where the primary cursor is positioned. |
| `%{cursorcolumn}`       | The position of the primary cursor inside the current line. |
| `%{selection}`          | The text selected by the primary cursor. |
| `%sh{cmd}`              | Executes `cmd` with the default shell and returns the command output, if any. |

archseer commented on toto.c:
```diff
@@ -0,0 +1 @@
+
```
Looks like a stray file was added

tdaron replied:
woops my bad

tdaron commented:
I think this is good for review (lints should be good now)

tdaron commented on helix-view/src/editor/variable_expansion.rs:
```diff
@@ -0,0 +1,159 @@
+use crate::Editor;
```
btw I didn't know why we are not using regex here. I think it could make the code way much clearer, but I kept it the way it was done in the original PR (my comment is about the whole file)

tdaron reviewed (commented):
.

tdaron reviewed (commented):
~~sorry for these useless comments I messed up with GitHub mobile~~

tdaron commented on helix-view/src/editor/variable_expansion.rs:
```diff
@@ -0,0 +1,159 @@
+                                                        command.output().await.map_err(|_| {
+                                                            anyhow::anyhow!(
+                                                                "Shell command failed: {body}"
+                                                            )
+                                                        })?;
+
+                                                    if output.status.success() {
+                                                        String::from_utf8(output.stdout).map_err(
+                                                            |_| {
+                                                                anyhow::anyhow!(
+                                                            "Process did not output valid UTF-8"
+                                                        )
+                                                            },
+                                                        )
+                                                    } else if output.stderr.is_empty() {
+                                                        Err(anyhow::anyhow!(
+                                                            "Shell command failed: {body}"
+                                                        ))
+                                                    } else {
+                                                        let stderr =
+                                                            String::from_utf8_lossy(&output.stderr);
+
+                                                        Err(anyhow::anyhow!("{stderr}"))
+                                                    }
+                                                })
+                                            });
+                                            o.push_str(&input[last_push_end..index]);
+                                            last_push_end = end + 1;
+
+                                            o.push_str(output?.trim());
+
+                                            break;
+                                        }
+                                    }
+                                } else if char == '{' {
+                                    right_bracket_remaining += 1;
+                                }
+                            }
+                        }
+                    }
```
btw I didn't know why we are not using regex here. I think it could make the code way much clearer, but I kept it the way it was done in the original PR (my comment is about the whole file)

RoloEdits commented:
Rather than tying this to `Shellwords` parsing, it makes more sense to me if it were to function as an ultra fancy `str::replace` that would return a `Cow<str>` that would then be passed to `Shellwords`, with `Shellwords::from(&args)`, with it none the wiser that anything happened.

Just like `str::replace`, it would copy to a new buffer and then replace when comes across the pattern(variable).

Perform all the replacements for the variables first and then evaluate the `%sh` if it needs to. Starting from inner `%sh` first and using the return from that in the parent `%sh` call.

I am working on #11149 to try to simplify the handling of args, and the way this is currently implemented would require a lot of changes for whichever would get merged second. Mostly with this pr's logic needing a complete rewrite.

tdaron commented:
Yup but isn't shellword parsing done at each keypress for completion,... ?

Would replacing before parsing using shellwords mean executing %sh at each keypress ?

tdaron commented:
Yup but isn't shellword parsing done at each keypress for completion,... ?

Would replacing before parsing using shellwords mean executing %sh at each keypress ? (I mean, each key typed after the %sh{...}

RoloEdits commented:
Yeah, currently this pr touches `shellwords`. Which I don't think it should.

tdaron commented:
It worked like this in the original PR but shellwords messed up with spaces inside variables expansion (eg: %sh{there are some spaces}) so I added a lil exception inside shellword to ignore spaces (actually, ignore everything) inside of a %{} but the expansion is only done when the command is ran

RoloEdits commented:
Ah, I see. With the changes I am making for the args I changed the `MappableCommand::Typeable.args` to a `String`
```rust
pub enum MappableCommand {
    Typable {
        name: String,
        args: String,
        doc: String,
    },
    Static {
        name: &'static str,
        fun: fn(cx: &mut Context),
        doc: &'static str,
    },
}
```

Instead of the current `master` `Vec<String>`. In this way I can just make an iterator over the string to parse out the proper arguments lazily. 

```rust
    pub fn execute(&self, cx: &mut Context) {
        match &self {
            Self::Typable { name, args, doc: _ } => {
                if let Some(command) = typed::TYPABLE_COMMAND_MAP.get(name.as_str()) {
                    let mut cx = compositor::Context {
                        editor: cx.editor,
                        jobs: cx.jobs,
                        scroll: None,
                    };

                    if let Err(err) =
                        (command.fun)(&mut cx, Args::from(args), PromptEvent::Validate)
                    {
                        cx.editor.set_error(format!("{err}"));
                    }
                }
            }
            Self::Static { fun, .. } => (fun)(cx),
        }
    }
```

If we were to treat the expanding as a fancy replace, we can just replace the `String` to make a `Cow<str>`, in case it doesnt have any variables to expand, and then pass that to `Args::from(&args)`

This could then be some rough form of the replacing:
```rust
fn expand_variables<'a>(editor: &Editor, args: &'a str) -> anyhow::Result<Cow<'a, str>> {
    let (view, doc) = current_ref!(editor);

    let mut expanded = String::with_capacity(args.len());
    let mut var = Tendril::new_const();
    let mut chars = args.chars().peekable();

    while let Some(c) = chars.next() {
        if c == '%' {
            if let Some('{') = chars.peek() {
                chars.next(); // consume '{'

                while let Some(&ch) = chars.peek() {
                    if ch == '}' {
                        chars.next(); // consume '}'
                        break;
                    }
                    var.push(ch);
                    chars.next();
                }

                match var.as_str() {
                    "basename" => {
                        let replacement = doc
                            .path()
                            .and_then(|it| it.file_name().and_then(|it| it.to_str()))
                            .unwrap();

                        expanded.push_str(replacement);
                    }
                    "filename" => {
                        let replacement = doc
                            .path()
                            .and_then(|path| path.parent())
                            .unwrap()
                            .to_str()
                            .unwrap();
                        expanded.push_str(replacement);
                    }
                    "dirname" => {
                        let replacement = doc
                            .path()
                            .and_then(|p| p.parent())
                            .and_then(std::path::Path::to_str)
                            .unwrap();
                        expanded.push_str(replacement);
                    }
                    "cwd" => {
                        let dir = helix_stdx::env::current_working_dir();
                        let replacement = dir.to_str().unwrap();
                        expanded.push_str(replacement);
                    }
                    "linenumber" => {
                        let replacement = (doc
                            .selection(view.id)
                            .primary()
                            .cursor_line(doc.text().slice(..))
                            + 1)
                        .to_string();

                        expanded.push_str(&replacement);
                    }
                    "selection" => {
                        let replacement = doc
                            .selection(view.id)
                            .primary()
                            .fragment(doc.text().slice(..));

                        expanded.push_str(&replacement);
                    }
                    unknown => bail!("unknown variable `{unknown}`"),
                }

                // Clear for potential further variables to expand.
                var.clear();
            } else {
                expanded.push(c);
            }
        } else {
            expanded.push(c);
        }
    }

    //... `%sh` stuff

    Ok(expanded.into())
}
```

to use it like :
```rust
match expand_variables(cx.editor, args) {
    Ok(args) => {
        if let Err(err) =
            (command.fun)(&mut cx, Args::from(&args), PromptEvent::Validate)
        {
            cx.editor.set_error(format!("{err}"));
        }
    }
    Err(err) => cx.editor.set_error(format!("{err}")),
};
```

RoloEdits commented:
A working concept with what I posted:
```toml
"A-," = [":sh echo `%{basename} %{filename} %{dirname} %{cwd} %{linenumber} %{selection}`"]
```
![image](https://github.com/user-attachments/assets/c087683a-b0ef-4bf7-bf0a-0b34e8a0f425)

tdaron commented:
Shall I update this PR to be rebased on yours or shall we directly integrate command expansions inside your refactoring ?

RoloEdits commented:
I'm not sure what changes will be proposed for my `Args` refactor pr, or if it might just be rejected altogether. For trying to coordinate multiple prs, I think we should ask for some help from the core maintainers.  Perhaps @archseer @the-mikedavis or @pascalkuthe has some direction they would like us to follow?

tdaron commented:
> Unless there is some other reason for parsing the variables in Shellwords, I think this separates things pretty well.

Don't completions need this ?

RoloEdits commented:
You mean for the `:echo` command introduced here? I believe this could be a static thing with the `CommandSigniture`.

tdaron commented:
Nope, I am talking about the code inside the command_mode to provide completion (but as I removed all completion specific code from the original PR I might have misunderstood how it works)

RoloEdits commented:
Yeah, I believe you can provide a `Vec<&str>` , `vec!["%{filename}"]`, that gets resolved by the completer. When a space is the final part of an input, for example, it should present the options given in the `CommandSigniture`. The issue is that with `%sh{}`, if you want to put variable inside it, I don't think the completer will work, as the last part of the input would be `}`.

```rust
        |editor: &Editor, input: &str| {
            let shellwords = Shellwords::from(input);
            let command = shellwords.command();

            if command.is_empty()
                || (shellwords.args().next().is_none() && !shellwords.ends_with_whitespace())
            {
                fuzzy_match(
                    input,
                    TYPABLE_COMMAND_LIST.iter().map(|command| command.name),
                    false,
                )
                .into_iter()
                .map(|(name, _)| (0.., name.into()))
                .collect()
            } else {
                // Otherwise, use the command's completer and the last shellword
                // as completion input.
                let (word, len) = shellwords
                    .args()
                    .last()
                    .map_or(("", 0), |last| (last, last.len()));

                TYPABLE_COMMAND_MAP
                    .get(command)
                    .map(|tc| tc.completer_for_argument_number(argument_number_of(&shellwords)))
                    .map_or_else(Vec::new, |completer| {
                        completer(editor, word)
                            .into_iter()
                            .map(|(range, file)| {
                                let file = shellwords::escape(file);

                                // offset ranges to input
                                let offset = input.len() - len;
                                let range = (range.start + offset)..;
                                (range, file)
                            })
                            .collect()
                    })
            }
        }, // completion
```

For instance this is how `:theme` and `:set` work, though those are not just a `vec![]` as they reference other aspects of how `helix` works. This would only be for a specific command though, like `:echo`, and wouldn't be a blanket way to get completions. But I also don't think that is wanted either.

RoloEdits commented:
And perhaps its actually **not** desired to parse in a way that respects whitespace? With the `Args` here being an iterator, you could filter all that aren't `}` and then `last` that? But when entering the arg you would have to make sure that its not like `%sh{echo hello}` as the last `}` would be touching, and without special handling, would be part of `last`'s output. 

If you wrote like `%sh{echo 'hello'}` it would be fine as the part in the quotes would be one arg, so `}` would be filtered out.

daedroza commented:
Is there a way we can increase the pop-up size? I'm attempting to get git log in a large pop-up instead of small one.

tdaron commented:
> Is there a way we can increase the pop-up size? I'm attempting to get git log in a large pop-up instead of small one.

Wouldn't more advanced solutions better suit your usecase ? (like a floating pane in zellij containing it (you can open it from helix) or something like that) 

But it should be possible, but not easy as this PR do not touch the TUI code part. We are using code already written and well integrated into helix. I don't know how hard it can be, and I think it's quite out of the scope of this PR.

danillos commented:
I did a test with `:echo %sh{git blame -L %{linenumber} %{filename}}`

and it is returning;

```
%sh{git blame -L 14 file.rb} 
```

So it didn't evaluate the shell script.

tdaron commented:
Yep, atm the PR does not handle completions inside of %sh, but I am waiting an answer from core maintainers about which direction we should follow with @RoloEdits as I don't want to waste my time writing code on this if it's going to be removed and re-written anyway haha

danillos commented:
> But you should get the desired behavior using `:sh git blame -L %{linenumber} %{filename}` I think, I didn't test

This was just an example. What I need is:

```
s = ':open %sh{~/.dotfiles/support/helix_command.rb switch_to_spec %{filename}}'
```

I moved back to the previous PR. It was working there.

daedroza commented:
> Yep, atm the PR does not handle completions inside of %sh, but I am waiting an answer from core maintainers about which direction we should follow with @RoloEdits as I don't want to waste my time writing code on this if it's going to be removed and re-written anyway haha

I don't think the core maintainers will allow this patch. Such kind of extensions will be covered with the Scheme language paradigm. I'd a similar issue where I wanted the current document's indentation size passed to formatter.

Most likely when https://github.com/helix-editor/helix/issues/10389 is fully ready probably.

Though I wish patches like yours are merged along with #11149

tdaron commented:
> > But you should get the desired behavior using `:sh git blame -L %{linenumber} %{filename}` I think, I didn't test
> 
> This was just an example. What I need is:
> 
> ```
> s = ':open %sh{~/.dotfiles/support/helix_command.rb switch_to_spec %{filename}}'
> ```
> 
> I moved back to the previous PR. It was working there.

I re-thought about it and yup the original PR did support it, but I didn't handled nested expansions inside the new shellwords parsing. Just fixed it and added tests for it. I hope both windows and macos support `echo` shell command as I used it in the tests :)

david-crespo commented:
I tried the latest commit (I'm on Mac) with the following command

```
C-B = ":echo %sh{git blame -c --date=short -L %{linenumber},+1 %{filename} | sed -E 's/[0-9]+).*//' | sed 's/(//g'}"
```

And got this:

<img width="946" alt="image" src="https://github.com/user-attachments/assets/4d6634aa-90b5-441d-80dd-f7d2549f3cf4">

danillos commented:
It doesn't work. You can use this command:

```
':sh echo "%sh{echo "%{filename}"}"'
```

It must print `your_file.rb`
But it is printing `%sh{echo you_file.rb}`

tdaron commented:
urgh..will look at this tonight

tdaron commented:
I think everything works 🍾.
 I successfully ran expansions from config.toml mappings as well as manually typed commands, and I was able to run other commands without any issue :)

hakan-demirli commented:
Is it possible to also add `cursorcolumn` as an expansion?

I am reimplementing my [helix-harpoon script](https://github.com/helix-editor/helix/issues/5167#issuecomment-2071437451) using command expansions and need to save the current cursor position to a file.

I understand if this is out of the scope of this PR.

andreystepanov commented:
In addition to `cursorcolumn` suggestion by @hakan-demirli, would it be possible to also include:

- `lang` (currently selected `:lang` for the buffer)
- `ext` (file extension, `md` part of `/some/path/README.md` file)

Both of these options almost always (not for `/some/path/config`) can be detected manually from `filename` path itself, but since this information is already available to the editor, it would be nice to expose it to `:sh`, `:pipe-to` and other commands.

Thanks

tdaron commented:
I added all of theses (lang, ext and cursorcolumn)

hakan-demirli commented:
Yes, it should start at 1 since it is consistent with the cursor position on the status line.

RoloEdits commented on helix-view/src/editor/variable_expansion.rs:
```diff
@@ -67,6 +67,22 @@ impl Editor {
+                                            + 1)
+                                        .to_string(),
+                                        "lang" => doc.language_name().unwrap_or("text").to_string(),
+                                        "ext" => match doc.relative_path() {
+                                            Some(path) => path
+                                                .to_string_lossy()
+                                                .split(".")
+                                                .last()
+                                                .unwrap_or("")
+                                                .to_string(),
+                                            _ => "".to_string(),
```
For here you should be able to use `Path::extension` in some form like:
```rust
doc.path()
.and_then(|p| p.extension())
.and_then(|e| e.to_str())
.unwrap_or_default();
```
Leaves any wonkiness that could happen to the rust implimentation. I think what you have now works, but you never know.

tdaron replied:
agreed, didn't know about this function, will update tomorrow

rcorre commented:
For reference, the original issue is #3134 (I had to dive through several linked PRs to find it).

fominok commented:
Hi, is there anything else needed to be done? Can't wait to reach Kakoune's territory with this

tdaron commented:
Hi !
There is some conflict with another PR (#11149) so we're waiting some direction to follow indications

baldwindavid commented:
@tdaron Thanks for this; working great for me. I do think it would be nice to have "relative" versions of `filename` and `dirname`.

As it is, you can't get at those without some extra work. For example, commands like this won't work as services like Github are going to need a relative rather than absolute path...

```
:sh gh browse %{filename}:%{linenumber}
```

tdaron commented:
you mean relative to your current shell directory or relative to the current project directory ? (If in my shell I am located at ..../cool-project/src with the git repo in cool-project, shall the paths be relative to cool-project or cool-project/src ?)

I didn't even know some cli tools required relative paths haha

david-crespo commented:
For the GH case (also something I want, see [hxblame.sh](https://github.com/david-crespo/dotfiles/blob/21a74982748b49ebd8d13b1a45ccc97ebf558b07/bin/hxblame.sh)) it would have to be relative to repo root. I can imagine also wanting it relative to working dir but I don’t know of a direct use case for it.

baldwindavid commented:
@tdaron I meant relative to the `cwd` variable which just happens to almost always correspond with a repo root for me. It would also be nice to have repo-specific variables. Regardless of the actual naming of the variables something like below...

```
| `%{rel_filename}`       | The path of the currently focused file relative to `cwd`. |
| `%{repo_dirname}`       | The absolute path of the current version control repository. |
| `%{repo_rel_filename}`  | The path of the currently focused file relative to `repo_dirname`. |
```

tdaron commented:
Hi ! 

Just added those :) You now have `filename:git_rel` as well as `filename:rel` ! :tada:

david-crespo commented:
Thank you, that's fantastic. I can now do in one line what I was previously doing with [a whole bash function](https://github.com/david-crespo/dotfiles/blob/6ba09db8314388297159dc7d8a8fd3ead69d6663/zshenv#L3-L15).

```toml
C-b = ":sh gh browse %{filename:git_rel}:%{linenumber} --commit=%sh{git rev-parse HEAD}"
```

(If the current commit hasn't been pushed yet this will 404, but that was a problem with my current function anyway.)

hakan-demirli commented:
@tdaron There is a regression on `cursorcolumn` feature.

Helix: 237cbe4bca46eed52efed39ed75eb44aaccbdde3
PR: 385ea7c07e9f742e68a148835ccf3cb8c43112d0

Cursor Column keeps increasing after each row:

![image](https://github.com/user-attachments/assets/976db881-23ca-4593-a803-5051cef0cee9)

It is only correct for the first row.

![image](https://github.com/user-attachments/assets/2fe318cd-47d8-4351-91aa-27a635439890)

7flash commented:
> Thank you, that's fantastic. I can now do in one line what I was previously doing with [a whole bash function](https://github.com/david-crespo/dotfiles/blob/6ba09db8314388297159dc7d8a8fd3ead69d6663/zshenv#L3-L15).
> 
> ```toml
> C-b = ":sh gh browse %{filename:git_rel}:%{linenumber} -c=%sh{git rev-parse HEAD}"
> ```
> 
> (If the current commit hasn't been pushed yet this will 404, but that was a problem with my current function anyway.)

How can I do this?

filename="$2"

^ do I need to recompile helix at this PR to have access to filename like in your bash script?

david-crespo commented:
@7flash Yes, you need to [build from source](https://docs.helix-editor.com/master/building-from-source.html). You could clone the fork directly and check out this branch, or (what I do) clone the main repo and add `git add remote` this fork as a remote. That way I can merge several PRs into my custom branch.

tdaron commented:
@hakan-demirli cursor column should be fixed :)

carterols commented:
Man, I'm excited to use this! Thanks for working on it. Hopefully it can get reviewed/merged soon.

tdaron commented:
Me too !
Just waiting for #11149 to be merged to rebase my work on it :)

baldwindavid commented:
@carterols - I've seen nothing to suggest it won't be merged _eventually_, but this PR has its roots in another PR from back in early last year so I'd suggest building from this branch or merging into your own if needing it. I've happily been doing so and it provides all the interoperability necessary to interact with the cli tooling I need (running tests under cursor, blame, open in Github, file explorer, interactive find and replace, etc). I know there's been a lot of discussion around an extension system, but I'm unsure whether I will actually want extensions anyway as opposed to vanilla Helix + best-of-breed cli tools. Right now we have neither in core, but pulling the branch and installing is easy enough for the only critical need I have.

tdaron commented:
Just merged master onto. Everything should be working fine, let me know if something broke

7flash commented:
Please merge, it will allow to execute all kinds of scripts, for exampl formatting currently opened buffer, assigned to Ctlr+F hotkey.

tdaron commented:
> Please merge, it will allow to execute all kinds of scripts, for exampl formatting currently opened buffer, assigned to Ctlr+F hotkey.

This PR cannot be merged until #11149 is done.
However, for your formating hotkey you can do it without it by selecting all the buffer, piping it into your formatter and replacing it with the result. (You can also directly use :format from helix which is simpler)

7flash commented:
> > Please merge, it will allow to execute all kinds of scripts, for exampl formatting currently opened buffer, assigned to Ctlr+F hotkey.
> 
> This PR cannot be merged until #11149 is done. However, for your formating hotkey you can do it without it by selecting all the buffer, piping it into your formatter and replacing it with the result. (You can also directly use :format from helix which is simpler)

Currently, I have hotkey in my helix config, however it applies to all files in the directory:

[keys.normal."+"]
f = ":run-shell-command deno fmt"

Trying to achieve a desired behavior, which does formatting only current file, I followed your suggestion. I've selected file into a buffer with Shift-% then pressed "|" to pipe a "deno fmt" command, but unfortunately it replaced my buffer with empty string.

I understand it might be because "deno fmt" does not expect a buffer but allows to pass a file path. In this case, I could write a program in Bun which expects a buffer and then passing it through formatter API as a temporary file, however the issue here that I would have to type full program path every time, to pipe through it, because it does not recognize my functions defined in ~/.zshrc.

darshanCommits commented:
Why not use the built-in formatter command.
It works on stdin, `deno fmt` also supports it.

Just use the command `deno fmt -` add to your configuration and it'll work seamlessly.

objectiveryan commented:
> > Please merge, it will allow to execute all kinds of scripts, for exampl formatting currently opened buffer, assigned to Ctlr+F hotkey.
> 
> This PR cannot be merged until #11149 is done. However, for your formating hotkey you can do it without it by selecting all the buffer, piping it into your formatter and replacing it with the result. (You can also directly use :format from helix which is simpler)

With https://github.com/helix-editor/helix/pull/4709 (which has already been merged into master but isn't yet in an official release) you can make key bindings like `"@:sh sometool <C-r>%<ret>"` to run sometool on the current file.

nestor-custodio commented:
Friendly reminder that some tooling behaves differently depending on which file you're running it against (sometimes based on the path, sometimes on the basename, sometimes on the extension, etc), and comments to the effect of "you can just pipe your file to whatever tool you want" or "just use the built-in `:format`/`:fmt`" make it seem like this PR is a nice-to-have when it in fact addresses a *need* many of us are currently having to work around.

To be clear, I'm not saying "don't help folks whose needs *can* be met with something like the built-in LSP formatter or a custom keybind/command" -- I'm saying please keep in mind any such suggestion may have effects beyond what the requester intended (depending on the tooling in question), and that in some cases there may not *be* any way to do (via current means) what this PR will allow.

Again: just a reminder to those helping and a heads-up to those asking for help.

nikitarevenco commented:
Since I imagine `%{basename}` and `%{dirname}` will be used quite often, can we have a shorthand for them like `%{b}` and `%{d}`? In vim that would be `%` and `%:p`

tdaron commented:
I do not think this is necessary, as those expansions will be almost only used inside a config file I think it's not a problem to have to write full variables names (and if they are used in the "interactive" command prompt a completion prompt could be created to be fast). Shortcuts would increase code complexity as well as documentation for a (too) small QOL improvement imo

nikitarevenco commented:
> I do not think this is necessary, as those expansions will be almost only used inside a config file I think it's not a problem to have to write full variables names (and if they are used in the "interactive" command prompt a completion prompt could be created to be fast). Shorthands would increase code complexity as well as documentation for a (too) small QOL improvement imo

I merged your PR into my personal fork as well as https://github.com/helix-editor/helix/pull/11285 and I've been using it for a few days. I gotta say, I'm seriously impressed by it

These two PRs together are really going to be game-changing. In my opinion, the **only** thing helix really needs is a file manager, above all else. https://github.com/helix-editor/helix/pull/11285 is great at providing the functionality to change files, while your PR is extremely useful for creating new files, deleting files and renaming files.

For example, I frequently run the following commands:
- Rename the current file: `:mv %{dirname}` 
- Change directory to current file's directory: `:cd %{dirname}`
- Create a new file in the current directory: `:o %{dirname}/file.txt`
- Create a new file in sibling directory: `:o %{dirname}/../lib/file.txt`

Helix doesn't even need a "full-blown" file manager because 95% of the functionality you would want will be available once this PR and https://github.com/helix-editor/helix/pull/11285 gets merged. Maybe in the future helix can have like a `:rm` to upgrade from 95% -> 99%

That's initially why I suggested to add shorthands for common variables. at least for me, I don't use it just in config files, but actively use the feature to navigate across the file system, completely erasing the need for anything more complex.

Regarding complexity of the change, I took a look at it and it turned out to be fairly straightforward. In `helix-view/src/editor/variable_expansion.rs` making the following changes on line 44, 51 and 77 is all that it takes to add the aliases:

```diff
- "basename" => doc
+ "basename" | "b" => doc
```

```diff
- "filename" => doc
+ "filename" | "f" => doc
```

```diff
- "dirname" => doc
+ "dirname" | "d" => doc
```

Introducing a completion is a great idea if you decide not to add this anyway, but its implementation would likely add a lot more complexity than adding aliases.

Even if a completion prompt was added, it would still be a few extra keystrokes for common operations (e.g. to tab through the list and then hit enter to select the one you want), so that's why I advocate for adding these. It's completely up to you though!

Having said that, I'm definitely looking forward to these two PRs getting into one of the next releases, as I think a lot of people who previously wouldn't use Helix will consider it more than before. Thank you!

tdaron commented:
Hey !
Thanks for this full explanation, I can now see the need for those shorthands, even if I think that having a completion system for variables expansions might solve this issue, allowing to write %{dirname} by only writing %{di then enter. But as this system is not implemented (I don't know if it's better to put it in another PR or in this one, and in both cases it will be a lot easier once #11149 got merged) I will add those shorthands :)

(When talking about code complexity I didn't meant implementation difficulty but rather a "keep it simple" approach to avoid adding unnecessary things)

tcoopman commented:
Thanks, loving this PR, now this is a lot easier: `":sh tmux send-keys -t '{right}' 'mix test %{filename:rel}:%{linenumber}' ENTER"`

One question/remark - given that there are some PRs to integrate `jujutsu` as a vcs, I'm wondering if it doesn't make more sense to rename `git_repo` and `git_rel` to just `repo` and `repo_rel`?

tdaron commented:
as a jujutsu user I couldn't agree more with you :)

tdaron commented:
As a jujutsu user I couldn't agree more with you :)

mkellyxp commented:
Ohhh I hope this gets merged...

Create a new file in the current directory: :o %{dirname}/file.txt
Delete current file :sh rm %{filename}

Would totally lesson my need for a file tree.  Much of my want for a file tree really comes down to file management.  Always having to type in the absolute path in my big app is always the biggest pain in helix

NikitaRevenco commented:
A neat usability improvement would be if the variable expansions generated filename completions.

Right now, they don't:

![image](https://github.com/user-attachments/assets/178d17a0-8020-4d10-b596-9f1b30a1ea2d)

But with a few additional lines of code, you can fix that:

![image](https://github.com/user-attachments/assets/2eadd8b2-6513-4eae-97bc-d28aec97c802)

In `helix-term/src/ui/mod.rs`, at line `440` after the initialization of the `is_tilde` variable, add the following:

```

        let path = editor
            .expand_variable_in_string(input)
            .map_or(input.to_string(), |z| z.into_owned());

        let path = path.as_str();
```

tdaron commented:
Just did :) 
I also added a check to avoid executing sh expansions at each keystrokes to provide path completions. (So they will not work with something such as `:e %sh{echo /}tmp/test.txt` for example).

And I also merged main, hope everything is fine but it should be, let me know if there are any issues :)

NikitaRevenco commented:
In the command prompt, if your cursor is directory after a `%`, and you type `{`, then it will add a closing pair automatically and place your cursor just before. By writing `%{` the actual effect will be `%{|}` where `|` is the cursor.

Thoughts on this?

rcorre commented:
Given we already have registers for inserting filenames/dirnames interactively, it seems odd to add a completion system. I suppose it's fine if it adds minimal complexity, but otherwise it seems like we should encourage using registers for interactive use and expansions for config use.

tdaron commented:
@NikitaRevenco I think adding such an "auto-close" feature in the prompt is a too small QOL improvement as it would need to edit parts of helix codebase this PR don't currently edit, and I don't think it's worth it

nestor-custodio commented:
If this works as-is, please just get it out in its current form so we can benefit from this feature on the next release. Any *improvements* to what this already does should then take the form of new (smaller and much more reviewable) PRs opened against master. Preventing this much-needed feature from possibly making the next release build over something as minor as autocompletion or filename suggestion niceties would be a shame.  🥺

david-crespo commented:
With #11950 merged, there is now a conflict in the docs page. Here is a suggestion for a merged version that puts the expansion at the end of the typable commands section, shorten the heading, and adds an entry to the TOC.

````md
# Commands

- [Typable commands](#typable-commands)
  - [Using variables](#using-variables-in-typable-commands)
- [Static commands](#static-commands)

## Typable commands

Typable commands are used from command mode and may take arguments. Command mode can be activated by pressing `:`. The built-in typable commands are:

{{#include ./generated/typable-cmd.md}}

### Using variables in typable commands

Helix provides several variables that can be used when typing commands or creating custom shortcuts. These variables are listed below:

| Variable                | Description |
| ---                     | ---                      |
| `%{basename}` or `%{b}` | The name and extension of the currently focused file. |
| `%{dirname}`  or `%{d}` | The absolute path of the parent directory of the currently focused file. |
| `%{cwd}`                | The absolute path of the current working directory of Helix. |
| `%{repo}`               | The absolute path of the VCS repository helix is opened in. Fallback to `cwd` if not inside a VCS repository|
| `%{filename}` or `%{f}` | The absolute path of the currently focused file. |
| `%{filename:rel}`       | The relative path of the file according to `cwd` (will give absolute path if the file is not a child of the current working directory) |
| `%{filename:repo_rel}`  | The relative path of the file according to `repo` (will give absolute path if the file is not a child of the VCS directory or the cwd) |
| `%{ext}`                | The extension of the current file |
| `%{lang}`               | The language of the current file   |
| `%{linenumber}`         | The line number where the primary cursor is positioned. |
| `%{cursorcolumn}`       | The position of the primary cursor inside the current line. |
| `%{selection}`          | The text selected by the primary cursor. |
| `%sh{cmd}`              | Executes `cmd` with the default shell and returns the command output, if any. |

#### Example

```toml
[keys.normal]
# Print blame info for the line where the main cursor is.
C-b = ":echo %sh{git blame -L %{linenumber} %{filename}}"
```
  
## Static Commands

Static commands take no arguments and can be bound to keys. Static commands can also be executed from the command picker (`<space>?`). The built-in static commands are:

{{#include ./generated/static-cmd.md}}
````

tdaron commented:
done :tada: !

RoloEdits commented:
Stated [here](https://github.com/helix-editor/helix/pull/12288#issuecomment-2555697324), it looks like #11149 is slated to be merged to main sometime after the next release.

Now that its certain that the changes there are going to be allowed, it might be worth trying to base the changes here on that branch to stay ahead of the changes. There should only be minor changes left so I don't think there will be any huge breaking changes, specially if the approach taken for expanding the variables is done as we discussed before, treating it as a fancy replace.

I'm willing to assist in any work that needs to be done to get it transitioned over. I have a branch I made that holds the concepts of what could be done [here](https://github.com/RoloEdits/helix/tree/variable-expansion) that could be used a base reference.

Working around a lot of command based things recently, #11149  #12288 , I have also started to look at #4423 as well, which would depend on the features added in those, as well as the changes here, specifically around handling arguments being passed.

An example of what I was thinking for that would be like:
```toml
[commands]
"wcd!" = [":write --force %{arg}", ":cd %sh{ %{arg} | path dirname }"]
# :wcd! %{repo}/sub/sub/file.txt
```

Getting this merged as is takes priority, but I'd also like not to heavily block future work, if at all possible. I dont think anything will, but just so the thought is out there, id thought i'd share.

As for some feedback on names, I wonder how people would feel about these changes?:
```text
basename -> file
dirname -> dir
filename -> path
linenumber -> line
cursorcolumn -> column
```

I still feel like `filename:rel` and `filename:repo_rel` don't feel too great, as far as names go, but also not sure of alternatives. I think composing is the way, for example in my `%{arg}`, `%{arg:1}`. For the `path` option, `rpath` could also work. But the equivalent for repo would be `rrepo`, which is also not what we'd want.

tdaron commented:
Yay some good news here 😁

Thanks for this large explanation, I'll be working on this next week

tdaron commented:
Everything should be working like previously but rebased on your changes. I'll see later to change the variable names as this is "minor", I just wanna be sure everything is working fine before

RoloEdits commented on helix-term/src/commands/typed.rs:
```diff
@@ -2529,6 +2424,17 @@ fn read(cx: &mut compositor::Context, args: &[Cow<str>], event: PromptEvent) ->
+        return Ok(());
+    }
+
+    let args = args.collect::<Vec<&str>>().join(" ");
```
```suggestion
    let args = args.raw();
```

This returns whatever was given to `Args::from`, so no need to the allocations.

tdaron replied:
fixed it. but i still need to clone it as editor's lifetime is bigger than args's one.

RoloEdits commented on helix-term/src/commands/typed.rs:
```diff
@@ -3184,67 +3099,66 @@ pub(super) fn command_mode(cx: &mut Context) {
 
-                if let Err(e) = (cmd.fun)(cx, &args[1..], event) {
-                    cx.editor.set_error(format!("{}", e));
+                    // Handle typable commands
+                    if let Some(cmd) = typed::TYPABLE_COMMAND_MAP.get(command) {
+                        if let Err(err) = (cmd.fun)(cx, shellwords.args(), event) {
+                            cx.editor.set_error(format!("{err}"));
+                        }
+                    } else if event == PromptEvent::Validate {
+                        cx.editor.set_error(format!("no such command: '{command}'"));
+                    }
```
Doesn't there need to be an expansion here?

tdaron replied:
```rust
        match expand_args(cx.editor, input.into(), true) {
                Ok(args) => {
                    let shellwords = Shellwords::from(args.as_ref());
```

shellwords are created after the expansion has been done

RoloEdits replied:
Ahh, I see it. The diff was a little confusing.

RoloEdits commented on helix-term/src/commands/expansions.rs:
```diff
@@ -0,0 +1,196 @@
+use helix_core::coords_at_pos;
+use helix_view::Editor;
+
+pub fn expand_args<'a>(
```
`expand_args` might lead to some confusion about what it does. My previous suggestion of an `variables.rs` just and `expand` should read `variables::expand(...)`, which to me is more clear what its doing, acting on only variables, not all `args` which it could suggest.

tdaron replied:
done :)

sanfilippopablo commented:
> As for some feedback on names, I wonder how people would feel about these changes?:
> 
> ```
> basename -> file
> dirname -> dir
> filename -> path
> linenumber -> line
> cursorcolumn -> column
> ```

I like those changes

NikitaRevenco commented:
> > As for some feedback on names, I wonder how people would feel about these changes?:
> > ```
> > basename -> file
> > dirname -> dir
> > filename -> path
> > linenumber -> line
> > cursorcolumn -> column
> > ```
> 
> I like those changes

I think these are more descriptive:

```
basename -> filename
dirname -> current-file-directory
filename -> current-file-path
linenumber -> line
cursorcolumn -> column
```

Reasons:
- `filename` is less ambiguous than `file`
- `dir` can be confused for current working directory. `current-file-directory` is more descriptive 
- `path` could also be confused for other things. for example, path of the current directory. ENV `$PATH`. `current-file-path`

`current-file-directory` and `current-file-path` are long, but we have shorthands so it's fine. Similar to how there is `:change-current-directory`. No one probably uses the full version, but if we have shorthands we may as well use longer, more descriptive names

NikitaRevenco commented:
But in all honesty, discussions over these details may actually delay this PR so it won't be in the next release. What about if we merge it as-is, and then make further enhancements down the line if necessary?

tdaron commented:
No matter what, this PR won't be included in the next release, as it depends on #11149, which will be merged after the next release.


I really like `line` and `column`. However I agree that `file` isn't descriptive enough, as well as `dir` (but dirname isn't that nice either).

I do think that:

filename
current_dir
path
line
column

Is great. But yes I think everyone will have a different opinion and those names could even change after this PR so it's not that important

RoloEdits commented:
Yes, if you look at https://github.com/helix-editor/helix/blob/changelog/CHANGELOG.md it says the first of the year. As not even one maintainer has left any reviews here yet, it wont be included for sure. Any big features will have to wait till next release cycle.

sanfilippopablo commented:
I don't think the names should be completely self-descriptive. After all, there's documentation with the full meaning if one's unsure about the meaning of any. And conciseness has value, although I'm not sure how much these substitutions will be used "live" instead of living in a keybinding/command in the config file.

RoloEdits commented:
#11149 was merged into master, so this can now be rebased off of master.

tdaron commented:
Hey !

Was trying to work on this but I noticed #11149 got unmerged from master inside of 217818681ea9bbc7f995c87f8794c46eeb012b1c

RoloEdits commented:
Ah, yes, I forgot to say something here: the maintainers felt more and more like they wanted a different API. The new work is being tracked here #12441.

Most of what the old work had done was liked, which is the bulk of the work, just wanted some more rules around escaping and more eager approach rather than a lazy one. Hopefully this can get merged soon. Sorry for the trouble on this.

tdaron commented:
No problem :)

RoloEdits commented:
Lucky I think the way that this PR is now setup it shouldn't be bad to merge. I will provide suggestions for conflict fixes when the time comes.

cor commented:
For those already using this PR, I wrote a small utility [gh-permalink](https://github.com/cor/gh-permalink) that copies the github permalink of your current line.

sanfilippopablo commented:
No way! I wrote the same thing a couple days ago, seems like we'll are wanting this PR for similar reasons:
```
#!/usr/bin/env fish

function get_git_provider_url --argument filename line_number
    if not git rev-parse --is-inside-work-tree >/dev/null 2>&1
        echo "Error: Not inside a git repository."
        return 1
    end

    set remote_url (git config --get remote.origin.url)
    if test -z "$remote_url"
        echo "Error: No remote origin found."
        return 1
    end

    if string match -q '*github.com*' $remote_url
        set provider github
    else if string match -q '*gitlab.com*' $remote_url
        set provider gitlab
    else
        echo "Error: Unsupported git provider."
        return 1
    end

    set repo_path (echo $remote_url | perl -nE 'say $2 if /\.com(:|\/)(.*)\.git/')
    set branch (git rev-parse --abbrev-ref HEAD)

    switch $provider
        case github
            echo "https://github.com/$repo_path/blob/$branch/$filename#L$line_number"
        case gitlab
            echo "https://gitlab.com/$repo_path/-/blob/$branch/$filename#L$line_number"
    end
end

if [ (count $argv) -ne 2 ]
    echo "Usage: get_git_url.fish <filename> <line_number>"
    return 1
end

get_git_provider_url $argv[1] $argv[2]
```

the-mikedavis commented:
I've merged down #12527 with a very minimal set of variables. An example for adding new variables can be found in https://github.com/the-mikedavis/helix/commit/967e2370c72f4cee39f72a31ef51cb72cf96bd78
